### PR TITLE
Add solc-cli module to fix 0.4.10 versions

### PR DESCRIPTION
**Pull request by sevenzing**

### CODE REVIEW

File: smart-contract-verifier/src/solidity/solc_cli.rs [hunk 1]
```diff
@@ -0,0 +1,493 @@
...
+    }
+
+    fn merge_libs(libraries: Libraries) -> BTreeMap<String, String> {
+        libraries
+            .libs
+            .into_iter()
+            .map(|(_, libs)| libs)
+            .reduce(|accum, item| accum.into_iter().chain(item).collect())
+            .unwrap_or_default()
+    }
```
Comment (rimrakhimov) [inline smart-contract-verifier/src/solidity/solc_cli.rs:60]: Maybe we may do something like this to make it more clear?

```suggestion
    fn merge_libs(libraries: Libraries) -> BTreeMap<String, String> {
        let mut result = BTreeMap::new();
        libraries
            .libs
            .into_iter()
            .for_each(|(_, libs)| result.extend(libs));
        result
    }
```

File: smart-contract-verifier/src/solidity/solc_cli.rs [hunk 2]
```diff
@@ -0,0 +1,493 @@
...
+mod types {
+    use super::serde_helpers;
+    use std::collections::{BTreeMap, HashMap};
+
+    use ethers_solc::{
+        artifacts::{Contract, Libraries},
+        error::SolcError,
+        CompilerInput, CompilerOutput,
+    };
+    use minicbor::encode::Write;
```
Comment (rimrakhimov) [inline smart-contract-verifier/src/solidity/solc_cli.rs:42]: That is not what you want to import

```suggestion

```

File: smart-contract-verifier/src/solidity/solc_cli.rs [hunk 3]
```diff
@@ -0,0 +1,493 @@
...
+                vec.push(libs.join(","))
+            }
+            vec
+        }
+    }
+
+    #[derive(Debug)]
+    pub struct InputFiles {
+        pub files_dir: TempDir,
+        pub file_names: Vec<String>,
```
Comment (rimrakhimov) [inline smart-contract-verifier/src/solidity/solc_cli.rs:104]: Why not to make it `PathBuf`?
```suggestion
        pub file_names: Vec<PathBuf>,
```

File: smart-contract-verifier/src/solidity/solc_cli.rs [hunk 4]
```diff
@@ -0,0 +1,493 @@
...
+
+            Ok(InputFiles {
+                files_dir,
+                file_names,
+            })
+        }
+    }
+
+    impl InputFiles {
+        pub fn build_string_args(&self) -> Result<&Vec<String>, SolcError> {
```
Comment (rimrakhimov) [inline smart-contract-verifier/src/solidity/solc_cli.rs:129]: Nit: maybe just `build`?

File: smart-contract-verifier/src/solidity/solc_cli.rs [hunk 5]
```diff
@@ -0,0 +1,493 @@
...
+        r#type: "".to_string(),
+        component: "".to_string(),
+        severity: Severity::Error,
+        error_code: None,
+        message,
+        formatted_message: None,
+    }
+}
+
+pub fn compile_using_cli(
```
Comment (rimrakhimov) [inline smart-contract-verifier/src/solidity/solc_cli.rs:221]: Nit: personally, I believe it would be better to rename that function to `compile`, make `pub(crate) mod solc_cli` inside `mod.rs`, and use it as `solc_cli::compile` from the `compilers.rs`

File: smart-contract-verifier/src/solidity/solc_cli.rs [hunk 6]
```diff
@@ -0,0 +1,493 @@
...
+    solc: &PathBuf,
+    input: &CompilerInput,
+) -> Result<CompilerOutput, SolcError> {
+    let output = {
+        let input_args = types::InputArgs::try_from(input)?;
+        let input_files = types::InputFiles::try_from(input)?;
+        Command::new(solc)
```
Comment (rimrakhimov) [inline smart-contract-verifier/src/solidity/solc_cli.rs:228]: I suggest to use `tokio::process::Command::new` and mark the function as `async`

**Review by rimrakhimov — APPROVED** [review]

File: smart-contract-verifier/src/solidity/solc_cli.rs (hunk 3 above)
Comment (sevenzing) [reply smart-contract-verifier/src/solidity/solc_cli.rs:104]: eventually in `build_string_args` I have to convert it to string anyway

File: smart-contract-verifier/src/solidity/solc_cli.rs (hunk 4 above)
Comment (sevenzing) [reply smart-contract-verifier/src/solidity/solc_cli.rs:129]: `input_files.build()`  looks like "build input files" for me

Comment (sevenzing) [reply smart-contract-verifier/src/solidity/solc_cli.rs:129]: ok maybe it make sense, not very important for me

File: smart-contract-verifier/src/solidity/solc_cli.rs (hunk 5 above)
Comment (sevenzing) [reply smart-contract-verifier/src/solidity/solc_cli.rs:221]: we have like 4 or 5 `compile` functions. don't you think it will confuse people?

File: smart-contract-verifier/src/solidity/solc_cli.rs (hunk 6 above)
Comment (sevenzing) [reply smart-contract-verifier/src/solidity/solc_cli.rs:228]: make sense

File: smart-contract-verifier/src/solidity/solc_cli.rs [hunk 7]
```diff
@@ -0,0 +1,494 @@
...
+                "bin-runtime": "60606040526000357c0100000000000000000000000000000000000000000000000000000000900463ffffffff168063481591811460505780636d4ce63c146070578063f3d33cf4146090575b6000565b34600057605a60b0565b6040518082815260200191505060405180910390f35b34600057607a60be565b6040518082815260200191505060405180910390f35b34600057609a60d3565b6040518082815260200191505060405180910390f35b60006414b230ce3890505b90565b600060c660d3565b60cc60b0565b0190505b90565b6000650712e7ae7c7190505b905600a165627a7a72305820a80a9599b36625e94a3eadfd5c31475a2c507be6d1a9fa9a35e9cd4c54bce1390029"
+            }
+        },
+        "version": "0.4.8+commit.60cc1668"
+    }
+    "#;
+
+    #[cfg(target_os = "linux")]
+    pub const LIST_JSON: &str = r#"
```
Comment (leviska) [inline smart-contract-verifier/src/solidity/solc_cli.rs:325]: imo we could minimize lists(?) (make them one line)

**Comment by leviska** [comment]
can you add a test with compile error (source code is broken) and execution error (for example, make `solc` file an empty file)?

I'd like to test that we're processing errors correctly

File: smart-contract-verifier/src/solidity/solc_cli.rs (hunk 7 above)
Comment (sevenzing) [reply smart-contract-verifier/src/solidity/solc_cli.rs:325]: I think it won't be readable for people who will try to understand why tests dont pass

**Review by leviska — APPROVED** [review]

File: smart-contract-verifier/src/solidity/solc_cli.rs (hunk 3 above)
Comment (rimrakhimov) [reply smart-contract-verifier/src/solidity/solc_cli.rs:101]: You can make `build_string_args` to return `Vec<PathBuf>` without any problem

File: smart-contract-verifier/src/solidity/solc_cli.rs (hunk 5 above)
Comment (rimrakhimov) [reply smart-contract-verifier/src/solidity/solc_cli.rs]: I actually believe that `compilers::compile` is excessive and сan just be directly inline inside `compilers::Compilers::compile`. So my idea was that after removal of `compilers::compile` we would have `solc_cli::compile` and `solc.compile` which does not seem actually confusing for me.
